pkg/controller/vsphere/session/test_ids_caching_client_test.go

```diff
@@ -0,0 +1,283 @@
+		Cardinality:     "SINGLE",
+		Name:            "test",
+	})
+	g.Expect(err).To(Succeed())
+	for _, tagName := range tagNames {
+		_, err := m.CreateTag(ctx, &tags.Tag{Name: tagName, CategoryID: testCategoryID})
+		g.Expect(err).To(Succeed())
+	}
+	for _, catName := range categoryNames {
+		_, err := m.CreateCategory(ctx, &tags.Category{
+			AssociableTypes: []string{"VirtualMachine"},
+			Cardinality:     "SINGLE",
+			Name:            catName,
+		})
+		g.Expect(err).To(Succeed())
+	}
+}
+
+func cleanupTagsAndCategories(ctx context.Context, m *CachingTagsManager, g Gomega) {
+	tagsList, err := m.ListTags(ctx)
+	g.Expect(err).To(Succeed())
+	for _, tagID := range tagsList {
+		g.Expect(m.DeleteTag(ctx, &tags.Tag{ID: tagID})).To(Succeed())
+	}
+	categoriesList, err := m.GetCategories(ctx)
+	g.Expect(err).To(Succeed())
+	for _, catID := range categoriesList {
+		g.Expect(m.DeleteCategory(ctx, &tags.Category{ID: catID.ID})).To(Succeed())
+	}
+}
+
+func purgeCache() {
+	sessionAnnotatedCache = map[string]*tagsAndCategoriesCache{}
+}
+
+func TestGetTag(t *testing.T) {
+	model, sessionObj, server := initSimulator(t)
+	defer model.Remove()
+	defer server.Close()
+
+	ctx := context.TODO()
+
+	t.Run("Tag Found", func(t *testing.T) {
+		g := NewWithT(t)
+		tagsToCreate := []string{"fooo", "bar", "baz", "fizz"}
+		categoriesToCreate := []string{"fizz"}
+
+		err := sessionObj.WithCachingTagsManager(context.TODO(), func(m *CachingTagsManager) error {
+			createTagsAndCategories(ctx, tagsToCreate, categoriesToCreate, m, g)
+			defer cleanupTagsAndCategories(ctx, m, g)
+			defer purgeCache()
+
+			g.Expect(len(sessionAnnotatedCache)).To(BeZero())
+
+			tag, err := m.GetTag(ctx, "baz")
+			g.Expect(err).To(Succeed())
+			g.Expect(tag).NotTo(BeNil())
+
+			// check cache filled
+			g.Expect(len(sessionAnnotatedCache)).To(BeEquivalentTo(1))
+			cachedTagId, found := getOrCreateSessionCache(sessionObj.sessionKey).tags.Get("baz")
+			g.Expect(found).To(BeTrue())
+			g.Expect(cachedTagId).Should(ContainSubstring("urn:"))
+
+			return nil
+		})
+		g.Expect(err).Should(Succeed())
+	})
+
+	t.Run("Tag deleted from vCenter after being cached", func(t *testing.T) {
+		g := NewWithT(t)
+		tagsToCreate := []string{"fooo", "bar", "baz", "fizz"}
+		categoriesToCreate := []string{"fizz"}
+
+		err := sessionObj.WithCachingTagsManager(context.TODO(), func(m *CachingTagsManager) error {
+			createTagsAndCategories(ctx, tagsToCreate, categoriesToCreate, m, g)
+			defer purgeCache()
+
+			g.Expect(len(sessionAnnotatedCache)).To(BeZero())
+
+			tag, err := m.GetTag(ctx, "fizz")
+			g.Expect(err).To(Succeed())
+			g.Expect(tag).NotTo(BeNil())
+
+			// check cache filled
+			g.Expect(len(sessionAnnotatedCache)).To(BeEquivalentTo(1))
+			cachedTagId, found := getOrCreateSessionCache(sessionObj.sessionKey).tags.Get("fizz")
+			g.Expect(found).To(BeTrue())
+			g.Expect(cachedTagId).Should(ContainSubstring("urn:"))
+
+			cleanupTagsAndCategories(ctx, m, g)
```

Comment:
in a real world scenario, is it possible that the tag would be deleted from the vCenter but would still exist in our cache for 12 hours?

Reply:
This particular scenario - no. Only ids are stored in cache, if tag would not be found in vCenter by cached value it will be invalidated.

https://github.com/openshift/machine-api-operator/pull/1057/files#diff-d4d1ee4b7e3a4119ecf48e49ba473e93b633e2cbb07a985c35721445b685d3f5R145

>and just for clarity, the "not found value" will exist in the cache until its TTL is over?

>what happens if the tag is not found and then the user creates it later, would they have to wait for 12 hours to pick it up or does is this scenario inaccurate in real world usage?

Speaking about not found value, yes, it will sit in the cache until TTL is over. There is no efficient way to check tag/category presence by name within vCenter, so, it is a tradeoff which we need to make. Restarting controller would help to reconcile the actual state, think it worth to document it somewhere.